### PR TITLE
faster `hypot`

### PR DESCRIPTION
**Pull request by oscardssmith**
unconditionally scale `hypot` rather than having separate branches for small/large/normal inputs.

### CODE REVIEW

**Comment by giordano** [comment]
Seems to break `hypot(floatmax(), floatmax())`

**Comment by oscardssmith** [comment]
good point. I need to clamp normalize.

**Comment by barucden** [comment]
Note that there is a typo (L721, `normalze`) at the current version.

File: base/math.jl [hunk 1]
```diff
@@ -711,6 +711,15 @@ julia> .√(1:4)
...
 sqrt(x)
 
+# get a scaling constant that you can multiply
+function normalize(x::T) where T<:IEEEFloat
+    UT = uinttype(T)
+    xi = reinterpret(UT, x)
+    ans = (~xi-(one(UT)<<significand_bits(T))) & exponent_mask(T)
+    reinterpret(T, ans)
+end
+normalze(x) = inv(x)
```
Comment (mikmoore) [inline base/math.jl:721]:
```suggestion
normalize(x) = inv(x)
```

File: base/math.jl [hunk 2]
```diff
@@ -785,17 +794,11 @@ function _hypot(x, y)
...
-    elseif ay < sqrt(floatmin(ax))
-        ax = ax/scale
-        ay = ay/scale
-    else
-        scale = oneunit(scale)
-    end
+    scale = normalize(ax)  #Rescaling constant
+    ax = ax*scale
+    ay = ay*scale
+    scale = inv(scale)
```
[the remark this answers is not in the document]
Comment (oscardssmith) [reply base/math.jl:800]: yeah I was having some trouble getting the bit manipulation working to do that well. Any suggestions?

[the remark this answers is not in the document]
Comment (mikmoore) [reply base/math.jl:800]: I was merely referring to
```julia
scale = inv(scale)
# much code
return h*scale*oneunit(axu)
```

But to your question, I've been playing with fast scaling in trying to improve `(/)(::ComplexF64,::ComplexF64)`. I don't have my actual implementation handy, but I was using something very close to this:
```julia
function inv_pow2(x::T) where T<:Base.IEEEFloat
	u = reinterpret(Unsigned, x)
	# floatmin(T) <= x <= inv(floatmin(T)) || error("out of range!") # ensure externally
	# u == u & Base.exponent_mask(T) || error("not a power of 2!") # ensure externally
	xinv = reinterpret(T, 2reinterpret(Unsigned, one(T)) - u)
	return xinv
end

```

**Pull request closed without merging**